backend/api/v1/views.py

```diff
@@ -37,121 +62,262 @@ def show_content(slug):
 
-    base_folder = current_app.config.get("USER_BASE_FOLDER")
+    if not os.path.exists(feedback_file):
+        abort(404)
 
-    notebook_folder = os.path.join(base_folder, student_slug, assignment_slug)
+    return send_file(feedback_file)
 
-    if os.path.exists(notebook_folder):
-        return redirect(os.path.join(request.host_url, "user-redirect/tree", assignment_slug))
-    elif content.git_url:
-        return redirect(content.git_url)
-    else:
-        abort(404)
 
+@blueprint.route("/feedback/<notebook_slug>", methods=["POST"])
+@is_grader
+def add_feedback(notebook_slug):
+    """
+    Add feedback for a given notebook
+    """
+    notebook = Notebook.query.filter_by(slug=notebook_slug).first_or_404()
+    submission_slug = notebook.submission_slug
+    assignment_slug = notebook.submission.content.assignment_slug
 
-@blueprint.route("/feedback/<slug>")
-@authenticated
-def get_feedback(user, slug):
-    submitted_notebook = grader_utils.get_submitted_notebook(slug)
+    folder = get_storage_folder(submission_slug, assignment_slug)
+    feedback_filename = os.path.join(folder, notebook.name + ".html")
 
-    if not submitted_notebook:
-        abort(404)
+    feedback_file = request.files.get("feedback")
 
-    student_slug = user["name"]
+    if not feedback_file:
+        abort(400)
 
-    if not submitted_notebook["student"] == student_slug:
-        abort(403)
+    score = request.form.get("score")
+
+    if score is None:
+        abort(400)
+
+    max_score = request.form.get("max_score")
 
-    assignment_slug = submitted_notebook["assignment_slug"]
-    notebook_name = submitted_notebook["name"]
+    if max_score is None:
+        abort(400)
 
-    feedback_path = grader_utils.get_feedback_path(student_slug, assignment_slug, notebook_name)
+    feedback_file.save(feedback_filename)
 
-    if not feedback_path:
+    notebook.score = score
+    notebook.max_score = max_score
+    notebook.graded = True
+    db.session.commit()
+
+    return jsonify({"status": "ok"})
+
+
+@blueprint.route("/notebook/<notebook_slug>")
+@is_grader
+def get_notebook(notebook_slug):
+    """
+    Return the notebook for the given notebook_slug if it is present.
+    """
+    notebook = Notebook.query.filter_by(slug=notebook_slug).first_or_404()
+    submission_slug = notebook.submission_slug
+    assignment_slug = notebook.submission.content.assignment_slug
+
+    folder = get_storage_folder(submission_slug, assignment_slug)
+    submission_file = os.path.join(folder, notebook.name)
+
+    if not os.path.exists(submission_file):
         abort(404)
 
-    return send_file(feedback_path)
+    return send_file(submission_file)
 
+
+@blueprint.route("/ungraded")
+@is_grader
+def get_ungraded_submissions():
+    """
+    Return information on the ungraded submissions.
+    """
+    # We need to do a python-side-filtering, as the property
+    # can not be calculated in SQL
+    submissions = Submission.query.all()
```

Comment:
Does that mean we have to query ALL submissions? I think from a scaling perspective this is not what we want. Can't we limit at least parts of it on the SQL side?

Reply:
So far, yes. It is possible, with [hybrid properties](https://docs.sqlalchemy.org/en/13/orm/extensions/hybrid.html) but this is definitely something for a follow-up (because it is much more complicated).

Reply:
Okay, I'm fine with it in the current state of the project. Thanks for your answer. :)